<a href="https://colab.research.google.com/github/FranItu/CODOJOTRACK1/blob/main/Week6/T1_Francisco_Iturralde.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#SEMANA 5 TAREA 3

##Header

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
import sklearn.metrics as skmt

In [ ]:
#Informacion de la base de datos
'''
Sex		nominal			M, F, and I (infant)
Length		continuous	mm	Longest shell measurement
Diameter	continuous	mm	perpendicular to length
Height		continuous	mm	with meat in shell
Whole weight	continuous	grams	whole abalone
Shucked weight	continuous	grams	weight of meat
Viscera weight	continuous	grams	gut weight (after bleeding)
Shell weight	continuous	grams	after being dried
Rings
'''

In [10]:
#header
hed = ['Sex','Length','Diameter','Height', 'Whole weight', 'Shucked weight','Viscera weight','Shell weight', 'Rings']

In [41]:
df = pd.read_csv('/content/drive/My Drive/Coding_Dojo/Bases_De_Datos/abalone.csv', header = None)
df.head()

0      1      2      3       4       5       6      7   8
0  M  0.455  0.365  0.095  0.5140  0.2245  0.1010  0.150  15
1  M  0.350  0.265  0.090  0.2255  0.0995  0.0485  0.070   7
2  F  0.530  0.420  0.135  0.6770  0.2565  0.1415  0.210   9
3  M  0.440  0.365  0.125  0.5160  0.2155  0.1140  0.155  10
4  I  0.330  0.255  0.080  0.2050  0.0895  0.0395  0.055   7

In [42]:
df.columns = hed
dforig = df

## ALUBÓN (Regresion)

In [ ]:
df['age'] = df.apply(lambda r: r.Rings + 1.5, axis = 1)
df2 = df
df2.pop('Rings')

In [18]:
dforig.head()

Sex  Length  Diameter  ...  Viscera weight  Shell weight   age
0   M   0.455     0.365  ...          0.1010         0.150  16.5
1   M   0.350     0.265  ...          0.0485         0.070   8.5
2   F   0.530     0.420  ...          0.1415         0.210  10.5
3   M   0.440     0.365  ...          0.1140         0.155  11.5
4   I   0.330     0.255  ...          0.0395         0.055   8.5

[5 rows x 9 columns]

In [9]:
y = df.loc[:, 'age'].values
y

array([16.5,  8.5, 10.5, ..., 10.5, 11.5, 13.5])

In [19]:
X = df
X.pop('age')
X.pop('Sex')

0       M
1       M
2       F
3       M
4       I
       ..
4172    F
4173    M
4174    M
4175    F
4176    M
Name: Sex, Length: 4177, dtype: object

In [20]:
X = X.loc[:,:].values
X

array([[0.455 , 0.365 , 0.095 , ..., 0.2245, 0.101 , 0.15  ],
       [0.35  , 0.265 , 0.09  , ..., 0.0995, 0.0485, 0.07  ],
       [0.53  , 0.42  , 0.135 , ..., 0.2565, 0.1415, 0.21  ],
       ...,
       [0.6   , 0.475 , 0.205 , ..., 0.5255, 0.2875, 0.308 ],
       [0.625 , 0.485 , 0.15  , ..., 0.531 , 0.261 , 0.296 ],
       [0.71  , 0.555 , 0.195 , ..., 0.9455, 0.3765, 0.495 ]])

In [22]:
X2 = df[['Length','Diameter','Height', 'Whole weight', 'Shucked weight','Viscera weight','Shell weight']]

In [31]:
reg = LinearRegression()  
reg.fit(X2, y)

LinearRegression()

In [32]:
print("Intercept: ", reg.intercept_)
print("Coefficients:")
list(zip(X2, reg.coef_))

Intercept:  4.485154181437299
Coefficients:


[('Length', -1.571897375474692),
 ('Diameter', 13.360916170642374),
 ('Height', 11.826072483837562),
 ('Whole weight', 9.247414445022951),
 ('Shucked weight', -20.213913185772917),
 ('Viscera weight', -9.829675481568211),
 ('Shell weight', 8.576242414916441)]

In [33]:
reg.score(X2, y)

0.5276299399919839

In [34]:
x_train, x_test, y_train, y_test = train_test_split(X2, y, test_size = 0.1, random_state = 100)

In [35]:
reg = LinearRegression()  
reg.fit(x_train, y_train)

LinearRegression()

In [36]:
print("Intercept: ", reg.intercept_)
print("Coefficients:")
list(zip(x_train, reg.coef_)) #la funcion zip devuelve tuplas con nombre de variable  + valor

Intercept:  4.40050588871493
Coefficients:


[('Length', -0.7939372209018942),
 ('Diameter', 13.079950504060662),
 ('Height', 10.819938564946945),
 ('Whole weight', 9.90824443856532),
 ('Shucked weight', -21.069827675036656),
 ('Viscera weight', -10.070523996583722),
 ('Shell weight', 7.471908856779839)]

In [38]:
reg.score(x_train, y_train)

0.5284120587407244

In [37]:
reg.score(x_test, y_test)

0.5184128399399383

## ALUBÓN (KNN)

In [43]:
df.head()

Sex  Length  Diameter  ...  Viscera weight  Shell weight  Rings
0   M   0.455     0.365  ...          0.1010         0.150     15
1   M   0.350     0.265  ...          0.0485         0.070      7
2   F   0.530     0.420  ...          0.1415         0.210      9
3   M   0.440     0.365  ...          0.1140         0.155     10
4   I   0.330     0.255  ...          0.0395         0.055      7

[5 rows x 9 columns]

In [44]:
y = df['Sex']
X = df[['Length','Diameter','Height', 'Whole weight', 'Shucked weight','Viscera weight','Shell weight', 'Rings']]

In [45]:
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(X, y)

KNeighborsClassifier(n_neighbors=3)

In [46]:
predictions = neigh.predict(X)
print(predictions)

['I' 'I' 'I' ... 'M' 'F' 'M']


In [47]:
print(skmt.accuracy_score(y, predictions))

0.7409624132152263


In [72]:
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 50)

In [73]:
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(x_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [74]:
predictions = neigh.predict(x_test)
print(predictions)

['F' 'F' 'I' 'M' 'I' 'I' 'F' 'I' 'I' 'F' 'M' 'F' 'I' 'F' 'M' 'I' 'F' 'I'
 'F' 'I' 'I' 'F' 'F' 'I' 'F' 'M' 'I' 'M' 'F' 'M' 'I' 'M' 'M' 'I' 'I' 'F'
 'F' 'F' 'I' 'I' 'M' 'I' 'I' 'I' 'I' 'M' 'F' 'I' 'I' 'F' 'M' 'F' 'F' 'M'
 'I' 'I' 'I' 'M' 'F' 'F' 'M' 'I' 'I' 'M' 'M' 'M' 'I' 'F' 'I' 'M' 'I' 'I'
 'F' 'F' 'F' 'F' 'M' 'M' 'M' 'M' 'I' 'M' 'F' 'M' 'F' 'F' 'I' 'I' 'I' 'F'
 'M' 'M' 'M' 'M' 'F' 'M' 'M' 'F' 'M' 'M' 'I' 'M' 'I' 'M' 'M' 'F' 'M' 'I'
 'I' 'F' 'F' 'M' 'F' 'I' 'M' 'I' 'F' 'M' 'I' 'M' 'M' 'M' 'F' 'F' 'F' 'M'
 'I' 'F' 'F' 'I' 'M' 'M' 'I' 'I' 'M' 'F' 'I' 'M' 'F' 'F' 'I' 'F' 'F' 'F'
 'M' 'M' 'M' 'M' 'I' 'I' 'F' 'M' 'F' 'I' 'I' 'M' 'F' 'I' 'I' 'F' 'I' 'M'
 'F' 'F' 'M' 'I' 'F' 'F' 'F' 'F' 'I' 'F' 'F' 'F' 'F' 'M' 'I' 'I' 'F' 'M'
 'F' 'F' 'M' 'F' 'F' 'F' 'M' 'F' 'F' 'F' 'M' 'F' 'I' 'I' 'I' 'M' 'M' 'F'
 'I' 'M' 'I' 'M' 'F' 'I' 'I' 'I' 'I' 'M' 'F' 'F' 'F' 'I' 'M' 'F' 'I' 'I'
 'I' 'F' 'I' 'F' 'F' 'M' 'M' 'M' 'F' 'M' 'M' 'F' 'M' 'F' 'F' 'I' 'M' 'I'
 'F' 'I' 'F' 'M' 'I' 'M' 'F' 'I' 'F' 'F' 'F' 'M' 'M

In [75]:
print(skmt.accuracy_score(y_test, predictions))

0.507177033492823


## Respuestas

**¿Cuál de KNN o regresión lineal pareció un mejor modelo cuando no usaste la división entrenar probar?**

El modelo de test train de la regresion tuvo menor desempeño que el modelo general. 0.527 vs 0.518

**¿Cuál de KNN o regresión lineal pareció un mejor modelo cuando usaste la división entrenar probar?**

El modelo de test train del KNN tuvo menor desempeño que el modelo general. 0.741 vs 0.507

**¿Hubo alguna ventaja en la regresión lineal en términos de la cantidad de código que tenías que escribir?**

Los dos son relativamente efectivos en codigo

**¿Hay alguna forma de mostrarle a alguien cuál de los dos modelos fue más efectivo?**

Si, mostrando el score de cada uno

**¿Crees que podrías haber mejorado KNN para mejorar la eficacia del modelo?**

Si, tomando un tamaño de test mas pequeño.
